doc/git-multipush.1
'\" t
.\"     Title: git-multpush
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 11/25/2013
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "GIT\-MULTPUSH" "1" "11/25/2013" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
git-multipush \- push a branch to multiple remotes in one command\&.
.SH "SYNOPSIS"
.sp
.nf
\fBgit multipush\fR [\fIOPTIONS\fR] [\fI<remote>\&...\fR] [\-\- \fIGIT_OPTIONS\fR]
\fBgit multipush\fR [\fIOPTIONS\fR] \fB\-s\fR \fI<remote>\fR [\&...]
\fBgit multipush\fR [\fIOPTIONS\fR] \fB\-\-unset\fR
\fBgit multipush\fR [\fIOPTIONS\fR] \fB\-g\fR
.fi
.SH "DESCRIPTION"
.sp
Particularly useful for people hosting on multiple git repo providers all at once\&. e\&.g\&., github\&.com, gitorious\&.org and repo\&.or\&.cz\&.
.SH "OPTIONS"
.PP
\fB\-v, \-\-verbose\fR
.RS 4
Print the git commands before executing them\&.
.RE
.PP
\fB\-e, \-\-error\fR
.RS 4
Fail immediately when any push fails\&. Otherwise, we fail after all pushes with the error code of the last failed push\&.
.RE
.PP
\fB\-n, \-\-dry\-run\fR
.RS 4
Don\(cqt run any of the git commands\&. Only print them, as in
\fB\-v\fR\&.
.RE
.PP
\fB\-b, \-\-branch\fR=\fI<branch>\fR
.RS 4
The branch to push\&. If none given, none are passed on to
\fBgit push\fR\&. If
\fI<branch>\fR
is given with no
\fI<remote>\*(Aqs and no \*(Aqmultipush\&.<branch>\&.remotes\fR
set, origin is used as the
\fI<remote>\fR\&.
.RE
.PP
\fB\-s, \-\-set\fR
.RS 4
Set
\fImultipush\&.<branch>\&.remotes\fR
to a comma\-separated list of the given
\fI<remotes>\fR\&. They will be used as the list of remotes to push to when none are passed explicitely\&.
.RE
.PP
\fB\-\-unset\fR
.RS 4
Unset
\fImultipush\&.<branch>\&.remotes\fR\&.
.RE
.PP
\fB\-g, \-\-get\fR
.RS 4
Get
\fImultipush\&.<branch>\&.remotes\fR
and print each remote on its own line\&.
.RE
.PP
\fB\-\-system\fR
.RS 4
Passed directly on to
\fIgit config\fR\&.
.RE
.PP
\fB\-\-global\fR
.RS 4
Passed directly on to
\fIgit config\fR\&.
.RE
.PP
\fB\-\-file\fR=\fI<file>\fR
.RS 4
Passed directly on to
\fIgit config\fR\&.
.RE
.PP
\fB\-z, \-\-null\fR
.RS 4
Only applies when
\fB\-g, \-\-get\fR
used\&. Print each remote with a null\-terminator instead of a newline\&. Useful with
\fBxargs \-0\fR
etc\&.
.RE
.PP
\fB\-\-version\fR
.RS 4
Print version info in the format
git multipush version $version\&.
.RE
.PP
<remote>\&.\&.\&.
.RS 4
The list of remotes to push to\&. None passed to
\fBgit push\fR
if none given\&.
.RE
.PP
\fIGIT_OPTIONS\fR
.RS 4
Options passed directly on to
\fBgit push\fR\&.
.RE
.SH "EXIT STATUS"
.sp
0 on success and non\-zero on failure\&.
.SH "AUTHOR"
.sp
Gavin Beatty <gavinbeatty@gmail\&.com>
.SH "RESOURCES"
.sp
Website: https://github\&.com/gavinbeatty/git\-multipush
.SH "REPORTING BUGS"
.sp
Please report all bugs and wishes to <gavinbeatty@gmail\&.com>
.SH "COPYING"
.sp
git\-multipush Copyright (C) 2010 Gavin Beatty, <gavinbeatty@gmail\&.com>
.sp
Free use of this software is granted under the terms of the GNU General Public License version 3, or at your option, any later version\&. (GPLv3+)


git-multipush.sh
#!/bin/sh

## git multipush Copyright 2010 Gavin Beatty <gavinbeatty@gmail.com>.
##
## This program is free software: you can redistribute it and/or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation, either version 3 of the License, or (at
## your option) any later version.
##
## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You can find the GNU General Public License at:
##   http://www.gnu.org/licenses/

set -e

# @VERSION@

version_print() {
    echo "git multipush version ${VERSION}"
}

SUBDIRECTORY_OK=Yes
OPTIONS_KEEPDASHDASH="yes"
OPTIONS_SPEC="\
git multipush [options] [<remote>..] [-- GIT_OPTIONS]
git multipush [options] -s <remote> [...]
git multipush [options] --unset
git multipush [options] -g
--
v,verbose   print each command as it is run
e,error     exit with error on the first git push error
n,dry-run   don't run any commands, just print them
b,branch=   directly passed on to git push as follows: git push <remote> <branch>
s,set       the given remotes are put in a comma-separated list in the config in multipush.<branch>.remotes
unset       unset multipush.<branch>.remotes
g,get       print each entry in multipush.<branch>.remotes on its own line
system      passed on to git config
global      passed on to git config
file=       passed on to git config
z,null      with -g|--get, print with \\000 terminator instead of \\n
d,debug     print debug info
version     print version info in 'git multipush version \$version' format"

. "$(git --exec-path)/git-sh-setup"

debug_run() {
    if test -n "$debug" ; then
        printf "debug:" >&2
        "$@" >&2
    fi
}
verbose_print() {
    if test -n "$verbose" ; then
        echo "$@" >&2
    fi
}
run() {
    if test -z "$dryrun" ; then
        "$@"
    fi
}

doit() {
    debug_run sq "$@"
    verbose_print "$@"
    run "$@"
}

sq() {
    git rev-parse --sq-quote "$@"
}

evalit() {
    e=0
    eval "$1" || e=$?
    if test "$e" -ne 0 ; then
        if test -n "$fail" ; then
            exit "$e"
        else
            exxit="$e"
        fi
    fi
    return 0
}

ref() {
    # doit can't be used because of bug... fix that
    git symbolic-ref "$1"
}

comma_assert() {
    for i in "$@" ; do
        if echo "$i" | grep -F -q ',' ; then
            die "Cannot add a default remote with a comma in its name"
        fi
    done
}

assert_HEAD() {
    debug_run echo ' using HEAD as branch'
    if ! git rev-parse --verify -q HEAD >/dev/null ; then
        die "Cannot work with detached HEAD and no <branch> given."
    fi
}

git_config() {
    if test -n "${file=}" ; then
        file=" $(sq --file "$file")"
    fi
    if test -n "${system=}" ; then
        system=" $(sq "$system")"
    fi
    if test -n "${global=}" ; then
        global=" $(sq "$global")"
    fi
    eval "doit git config${system}${global}${file} $(sq "$@")"
}

main() {
    dryrun=""
    verbose=""
    fail=""
    debug=""
    branch=""
    set_opt=""
    unset_opt=""
    get_opt=""
    term="\n"
    while test $# -ne 0 ; do
        case "$1" in
        -v|--verbose)
            verbose="true"
            ;;
        -e|--error)
            fail="true"
            ;;
        -n|--dry-run)
            dryrun="true"
            verbose="true"
            ;;
        -b|--branch)
            branch="$2"
            shift
            ;;
        -s|--set)
            set_opt="true"
            ;;
        --unset)
            unset_opt="true"
            ;;
        -g|--get)
            get_opt="true"
            ;;
        --system)
            system="--system"
            ;;
        --global)
            global="--global"
            ;;
        --file)
            file="$2"
            shift
            ;;
        -z|--null)
            term="\000"
            ;;
        -d|--debug)
            debug="true"
            verbose=""
            ;;
        --version)
            version_print
            exit 0
            ;;
        --)
            shift
            break
            ;;
        esac
        shift
    done

    if test -n "$set_opt" ; then
        if test $# -eq 0 ; then
            die "Must give at least one <remote> with -s|--set"
        fi
        remote_commas="$1"
        shift
        comma_assert "$remote_commas"
        for rem in "$@" ; do
            comma_assert "$rem"
            remote_commas="${remote_commas},${rem}"
        done
        if test -z "$branch" ; then
            assert_HEAD
            branch="$(ref "HEAD" | sed -e 's|^refs/heads/||')"
        fi
        git_config "multipush.${branch}.remotes" "$remote_commas"
        exit 0
    elif test -n "$unset_opt" ; then
        if test $# -ne 0 ; then
            die "Unexpected arguments given with --unset"
        fi
        if test -z "$branch" ; then
            assert_HEAD
            branch="$(ref "HEAD" | sed -e 's|^refs/heads/||')"
        fi
        git_config --unset "multipush.${branch}.remotes"
        exit 0
    elif test -n "$get_opt" ; then
        if test -z "$branch" ; then
            assert_HEAD
            branch="$(ref "HEAD" | sed -e 's|^refs/heads/||')"
        fi
        remotes="$(git_config "multipush.${branch}.remotes" | sed -e 's/,/ /g')"
        for rem in $remotes ; do
            printf "${rem}${term}"
        done
        exit 0
    fi

    while test $# -gt 0 ; do
        if test x"$1" = x"--" ; then
            shift
            break
        fi
        remotes="${remotes- }$(sq "$1")"
        shift
    done
    while test $# -gt 0 ; do
        git_opts="${git_opts- }$(sq "$1")"
        shift
    done

    eval set -- "${remotes-}"

    exxit=0
    if test $# -gt 0 ; then
        for rem in "$@" ; do
            if test -z "${branch-}" ; then
                evalit "doit git push${git_opts-} $(sq "$rem")"
            else
                evalit "doit git push${git_opts-} $(sq "$rem") $(sq "$branch")"
            fi
        done
    else
        e=0
        if test -z "$branch" ; then
            assert_HEAD
            hbranch="$(ref "HEAD" | sed -e 's|^refs/heads/||')"
        else
            hbranch="$branch"
        fi
        remote_commas="$(git_config "multipush.${hbranch}.remotes" || e=$?)"
        debug_run echo " multipush.${hbranch}.remotes =$(sq "$remote_commas")"
        if test "$e" -eq 0 && test -n "${remote_commas-}" ; then
            remotes="$(echo "$remote_commas" | sed -e 's/,/ /g')"
            for rem in $remotes ; do
                evalit "doit git push${git_opts-}$(sq "$rem")$(sq "$hbranch")"
            done
        else
            if test -z "${branch-}" ; then
                evalit "doit git push${git_opts-}"
            else
                evalit "doit git push${git_opts-} origin$(sq "$hbranch")"
            fi
        fi
    fi
    exit "$exxit"
}

trap "echo \"caught SIGINT\" ; exit 1 ;" INT
trap "echo \"caught SIGTERM\" ; exit 1 ;" TERM
trap "echo \"caught SIGHUP\" ; exit 1 ;" HUP

main "$@"

